### PR TITLE
Lecture 1 solutions

### PR DESCRIPTION
**Pull request by evlogiy**
### Solutions for Lecture 1

cc @chshersh @vrom911

### CODE REVIEW

File: src/Lecture1.hs [hunk 1]
```diff
@@ -50,7 +51,8 @@ Explanation: @sumOfSquares 3 4@ should be equal to @9 + 16@ and this
 is 25.
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-sumOfSquares x y = error "TODO!"
+sumOfSquares :: Num a => a -> a -> a
```
Comment (chshersh) [inline src/Lecture1.hs]: `Num` is explained in the following lectures, so you are one step ahead 😸
In here `Int -> Int -> Int` or `Integer -> Integer -> Integer` would also work, as a more specific type 🙂

File: src/Lecture1.hs [hunk 2]
```diff
@@ -63,7 +65,8 @@ sumOfSquares x y = error "TODO!"
 
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Int -> Int
+lastDigit n = mod (abs n) 10
```
Comment (chshersh) [inline src/Lecture1.hs]: Absolutely correct 👍🏻 
Alternatively, you can use the infix form of a function by putting it inside backticks :slightly_smiling_face: 
```suggestion
lastDigit n = abs n `mod` 10
```

File: src/Lecture1.hs [hunk 3]
```diff
@@ -77,7 +80,13 @@ and 1 is the smallest, and 7 - 1 = 6.
...
 function.
 -}
-minmax x y z = error "TODO"
+
+
+minmax :: Int -> Int -> Int -> Int
+minmax x y z = max a $ max b c
+    where a = abs (x - y)
+          b = abs (x - z)
+          c = abs (z - y)
```
Comment (chshersh) [inline src/Lecture1.hs:89]: Very clever solution! 🤯

File: src/Lecture1.hs [hunk 4]
```diff
@@ -119,4 +138,19 @@ and lower than 6 elements (4, 5, 6, 7, 8 and 9).
...
+
+lowerAndGreater :: Int -> [Int] -> [Char]
+lowerAndGreater number arr = printOut number greaterNum lowerNum
+    where lowerAndGreaterArr :: Int -> [Int] -> [Int]
+          lowerAndGreaterArr _ []     = []
+          lowerAndGreaterArr n [x]    | x > n = [1] | x < n = [-1] | otherwise = [0]
+          lowerAndGreaterArr n (x:xs) = lowerAndGreaterArr n [x]  ++ lowerAndGreaterArr n xs
+
+          numTimesFound :: Eq a => a -> [a] -> Int
+          numTimesFound x = length . filter (== x)
```
Comment (chshersh) [inline src/Lecture1.hs]: That's correct! 💯 

However, with this solution, you will walk through the list twice.
With the recursive solution and manual check on every element in the list on both conditions at the same time (without `filter` function), you would be able to write it more performant with only one traversal of the list 🙂

**Review by chshersh — APPROVED** [review]
Congrats! The first homework is complete 👏🏼 
Hope you enjoyed the course so far. Keep up the great work, and feel free to ask any questions if you have 👌🏼

**Comment by evlogiy** [comment]
Thanks for review, @chshersh! ❤️

**Pull request merged**